File: web/src/dtos/Shipment.ts
import Common from "../utility/Common";
import { DeliveryPlanRes } from "./DeliveryPlan";
import { OrderRes } from "./Order";
import { PagedReq } from "./PagedReq";
import { ShipmentAddressRes } from "./ShipmentAddress";
import { ShipmentItemReqEdit, ShipmentItemRes } from "./ShipmentItem";

export interface ShipmentRes {
  shipmentId?: number;
  orderId?: number;
  order?: OrderRes;
  trackingNumber?: string;
  deliveryPlanId?: number;
  deliveryPlan?: DeliveryPlanRes;
  bookingDate?: Date;
  deliveryDate?: Date;
  shipmentAddressId?: number;
  shipmentAddress?: ShipmentAddressRes;

  shipmentItems?: ShipmentItemRes[];
}

export class ShipmentReqCreate {
  orderId?: number = 0;
  trackingNumber?: string = "";
  deliveryPlanId?: number = 0;
  shipmentItems?: ShipmentItemReqEdit[];
}

export class ShipmentReqEdit {
  orderId?: number = 0;
  trackingNumber?: string = "";
  deliveryPlanId?: number = 0;
  deliveryDate?: Date = new Date();
}

export class ShipmentReqSearch extends PagedReq {
  orderId?: number;
  constructor(
    {
      pageNumber = 1,
      pageSize = Common.DEFAULT_PAGE_SIZE,
      orderBy,
      searchText = "",
    }: PagedReq,
    { orderId = 0 }
  ) {
    super({
      pageNumber: pageNumber,
      pageSize: pageSize,
      orderBy: orderBy,
      searchText: searchText,
    });
    this.orderId = orderId;
  }
}

File: data/Dtos/Shipment.cs
using data.Entities;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using data.Utility.Paging;

namespace data.Dtos
{
    public class ShipmentRes
    {
        public int ShipmentId { get; set; }
        public int OrderId { get; set; }
        public Order? Order { get; set; }
        public string? TrackingNumber { get; set; }
        public int DeliveryPlanId { get; set; }
        public DeliveryPlan? DeliveryPlan { get; set; }
        public DateTime BookingDate { get; set; } = DateTime.UtcNow;
        public DateTime? DeliveryDate { get; set; }
        public int ShipmentAddressId { get; set; }
        public ShipmentAddress? ShipmentAddress { get; set; }
        public ICollection<ShipmentItem>? ShipmentItems { get; set; }
    }

    public class ShipmentReqEdit
    {
        [Required]
        public int OrderId { get; set; }
        public string? TrackingNumber { get; set; }
        [Required]
        public int DeliveryPlanId { get; set; }
        public DateTime BookingDate { get; set; } = DateTime.UtcNow;
        public DateTime? DeliveryDate { get; set; }
        
    }

    public class ShipmentReqSearch : PagedRequestDto
    {
        public int OrderId { get; set; }
    }

}
